### PR TITLE
More work on api/yarpcerrors

### PR DESCRIPTION
**Pull request by peter-edge**
This PR does a few things:

- Explicitly document `IsYARPCError(err)` to be equivalent to `ErrorCode(err) == CodeOK`. I have a bunch of TODOs on #1098 that comment that this is not explicit, so there is a mismatch between using `IsYARPCError` and expecting an error code from `ErrorCode` that is not `CodeOK`.
- Add the function `ToYARPCError`. This is a convenience function that makes sure you have a YARPC error, this will remove a lot of logic in transport implementations.
- Add validation for user-defined names to only be lowercase letters from a-z and dashes.
- Add some documentation noting that all `yarpcErrors` created cannot have `Code` set to `CodeOK`.

### CODE REVIEW

File: api/yarpcerrors/errors.go [hunk 1]
```diff
@@ -25,20 +25,32 @@ import (
 	"fmt"
 )
 
-// IsYARPCError returns true if the given error is a non-nil YARPC error.
+// IsYARPCError is a convenience function that returns true if the given error
+// is a non-nil YARPC error.
+//
+// This is equivalent to yarpcerrors.ErrorCode(err) == yarpcerrors.CodeOK.
```
Comment (kriskowal) [inline api/yarpcerrors/errors.go]: !=

File: api/yarpcerrors/name.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+	// I think that name is a global reference to a slice effectively, so len(name) has some overheade
+	// https://stackoverflow.com/questions/26634554/go-multiple-len-calls-vs-performance
+	// https://blog.golang.org/strings
+	lenNameMinusOne := len(name) - 1
+	for i, b := range name {
+		if (i == 0 || i == lenNameMinusOne) && b == '-' {
+			return InternalErrorf("invalid error name, must only start or end with lowercase letters: %s", name)
+		}
+		if _, ok := _validNameRunes[b]; !ok {
```
Comment (kriskowal) [inline api/yarpcerrors/name.go]: This would be more concise as `rune >= 'a' && rune <= 'z' || rune == '-'`

File: api/yarpcerrors/errors.go (hunk 1 above)
Comment (peter-edge) [inline api/yarpcerrors/errors.go]: Done

File: api/yarpcerrors/name.go (hunk 2 above)
Comment (peter-edge) [inline api/yarpcerrors/name.go]: Done

**Pull request merged**